### PR TITLE
Removes virtual-specific subtypes of drinking glass + adds a new fishing map

### PR DESCRIPTION
**Pull request by jlsnow301**
## About The Pull Request
Changed how loot signals work with bitrunning entirely, which allows map creators to attach functionality to objects without creating subtypes of the item

As a bonus I added a fishing minigame map which uses it
## Why It's Good For The Game
It's a messy solution to need to create subtypes just for the one-off map that needs them
## Changelog
:cl:
add: Added a new fishing map to bitrunning.
add: You are no longer limited to pina coladas on the beach bar domain. Cheers!
/:cl:

### CODE REVIEW

File: code/modules/bitrunning/virtual_domain/domains/beach_bar.dm [hunk 1]
```diff
@@ -13,10 +13,18 @@
...
 	. = ..()
 
-	AddComponent(/datum/component/bitrunning_points, \
-		signal_type = COMSIG_GLASS_DRANK, \
-		points_per_signal = 0.5, \
-	)
+	for(var/obj/item/reagent_containers/cup/glass/drink in created_atoms)
+		if(istype(drink, /obj/item/reagent_containers/cup/glass/drinkingglass/filled))
+			drink.list_reagents += list(/datum/reagent/consumable/ethanol/pina_colada = 30)
+			drink.add_initial_reagents()
```
Comment (MrMelbert) [inline code/modules/bitrunning/virtual_domain/domains/beach_bar.dm:22]: If these drinking glasses have already been initialized, then you shouldn't call `add_initial_reagents` a second time. 

You should just be doing `drink.reagents.add_reagent(/datum/reagent/consumable/ethanol/pina_colada, 30)`.

File: code/modules/bitrunning/virtual_domain/domains/beach_bar.dm [hunk 2]
```diff
@@ -13,10 +13,18 @@
 	desc = "Whose drink is this? Not yours, that's for sure. Well, it's not like they're going to miss it."
 	list_reagents = list(/datum/reagent/consumable/ethanol/pina_colada = 30)
```
Comment (MrMelbert) [inline code/modules/bitrunning/virtual_domain/domains/beach_bar.dm:14]: You didn't remove `/obj/item/reagent_containers/cup/glass/drinkingglass/filled/virtual_domain`

File: code/modules/bitrunning/components/bitrunning_points.dm [hunk 3]
```diff
@@ -1,46 +1,41 @@
...
+	/// The amount required to spawn a crate
+	var/points_goal = 10
+	/// A special condition limits this from spawning a crate
+	var/points_received = 0
+	/// Finished the special condition
+	var/revealed = FALSE
+
+/datum/component/bitrunning_points/Initialize(datum/lazy_template/virtual_domain/domain)
+	. = ..()
+	if(isturf(parent))
```
Comment (MrMelbert) [inline code/modules/bitrunning/components/bitrunning_points.dm:12]: The doc comment says `Attaches to a turf` but this fails if parent is a turf

File: code/modules/bitrunning/virtual_domain/domains/beach_bar.dm [hunk 4]
```diff
@@ -13,10 +13,18 @@
...
-/obj/item/reagent_containers/cup/glass/drinkingglass/filled/virtual_domain/Initialize(mapload, vol)
+/datum/lazy_template/virtual_domain/beach_bar/setup_domain(list/created_atoms)
 	. = ..()
 
-	AddComponent(/datum/component/bitrunning_points, \
-		signal_type = COMSIG_GLASS_DRANK, \
-		points_per_signal = 0.5, \
-	)
+	for(var/obj/item/reagent_containers/cup/glass/drink in created_atoms)
+		if(istype(drink, /obj/item/reagent_containers/cup/glass/drinkingglass/filled))
```
Comment (MrMelbert) [inline code/modules/bitrunning/virtual_domain/domains/beach_bar.dm:20]: This feels icky, the filled glass subtype is meant to just be a helper for "glasses that spawn filled". But you're using it as a pseudo mapping helper.

You should map in `/obj/item/reagent_containers/cup/glass/drinkingglass/filled/pina_colada` into the map instead.

File: code/modules/bitrunning/virtual_domain/domains/beach_bar.dm (hunk 2 above)
Comment (jlsnow301) [reply code/modules/bitrunning/virtual_domain/domains/beach_bar.dm]: i want to blame merge artifacts here but I don't even think I can tbh

File: code/modules/bitrunning/virtual_domain/domains/beach_bar.dm [hunk 5]
```diff
@@ -8,15 +8,14 @@
...
-		points_per_signal = 0.5, \
-	)
+	for(var/obj/item/reagent_containers/cup/glass/drink in created_atoms)
+		RegisterSignal(drink, COMSIG_GLASS_DRANK, PROC_REF(on_drink_drank))
+
+/// Eventually reveal the cache
+/datum/lazy_template/virtual_domain/beach_bar/proc/on_drink_drank(datum/source)
+	SIGNAL_HANDLER
+
+	SEND_SIGNAL(src, COMSIG_BITRUNNER_GOAL_POINT, 0.5)
```
Comment (Jacquerel) [inline code/modules/bitrunning/virtual_domain/domains/beach_bar.dm:21]: We're back to "sending the same signal from a bunch of places" again, can this be a proc on `lazy_template/virtual_domain`?

**Review by Jacquerel — CHANGES_REQUESTED** [review]

**Pull request merged**